src/ml_board/frontend/Dashboard/src/redux/charts/chartsSlice.ts

```diff
@@ -0,0 +1,90 @@
+import { createEntityAdapter, createSlice, EntityState, PayloadAction } from "@reduxjs/toolkit";
+import { ChartUpdate } from "../../worker_socket/event_handlers/EvaluationResultHandler";
+import { RootState } from "../store";
+
+interface Experiment {
+    exp_id: number,
+    data: Array<number>,
+}
+
+export interface Chart {
+    chart_id: string,
+    x_axis: Array<number>,
+    experiments: EntityState<Experiment>,
+}
+
+const chartsAdapter = createEntityAdapter<Chart>({
+    selectId: ({ chart_id }: Chart) => chart_id,
+});
+
+const experimentsAdapter = createEntityAdapter<Experiment>({
+    selectId: ({ exp_id }: Experiment) => exp_id,
+});
+
+const initialState: EntityState<Chart> = chartsAdapter.getInitialState({});
+
+export const chartsSlice = createSlice({
+    name: 'charts',
+    initialState,
+    reducers: {
+        upsertCharts(state, { payload }: PayloadAction<ChartUpdate[]>) {
```

Comment:
I see a lot of `.includes(.)` function calls below. Are those dictionary lookups with O(1). If it is O(n) it would get super slow over time.

Reply:
according to [this](https://tc39.es/ecma262/#sec-array.prototype.includes) it's O(n) but we can sort at insertion time and then index to get them in O(1).

But I'm working on this code in #155 so let me edit it there better!

Reply:
ok please add an issue for this for tracking and which we will then close with PR #155.

Reply:
Done!